### PR TITLE
Add `logei_candidate_func` and make it default when available

### LINKED ISSUE #4546: Use `LogExpectedImprovement` for `BoTorchSampler`
Currently, `BoTorchSampler` uses `qExpectedImprovement` acquisition function. This relies on quasi-Monte Carlo sampling to compute the acquisition function, which is not accurate when the probability of improvement is very low. 

Quasi-Monte Carlo methods are last-resort methods for computing acquisition functions that are not analytically computable. This is the case when `q >= 2`, i.e. in batch optimization, but Optuna does not do batch optimization, so the expected improvement can be computed analytically.

BoTorch has a better acquisition function class `botorch.acquisition.analytic.LogExpectedImprovement` (defined as `log` of  `ExpectedImprovement`). Maximizing this function is mathematically equivalent to maximizing `ExpectedImprovement`, but numerically maximizing `ExpectedImprovement` is difficult because it tends to be very small for the vast area of the search space (and also susceptible to precision loss). Instead, `LogExpectedImprovement` can provide more meaningful gradient information for the optimizer.

### PR DESCRIPTION
**Pull request by contramundum53**
## Motivation
Resolve #4546.

The rationale for this improvement and the official benchmarks are available [here](https://github.com/pytorch/botorch/pull/1565).

## Description of the changes
* Add `logei_candidate_func`.
* Use `logei_candidate_func` as default when there is no constraint. ~and `LogExpectedImprovement` is available.~
* Modify tests accordingly.

Note that `LogExpectedImprovement` has the following limitations:
* Constraints are not available; supporting constraints needs quasi-Monte Carlo sampling.
* `LogExpectedImprovement` is only available in `botorch >=0.8.1`.

In those cases, this PR simply falls back to `qei_candidate_func`.

### CODE REVIEW

**Comment by contramundum53** [comment]
@toshihikoyanase @not522 Could you review this PR? If you are busy, please feel free to unassign yourself.

**Comment by contramundum53** [comment]
`X_pending` also seems to be unavailable in `LogExpectedImprovement`. Maybe we should set `consider_running_trial=False` by default?

File: tests/integration_tests/test_botorch.py [hunk 1]
```diff
@@ -97,6 +102,13 @@ def candidates_func(
     ],
 )
 def test_botorch_specify_candidates_func(candidates_func: Any, n_objectives: int) -> None:
+    if candidates_func == integration.botorch.logei_candidates_func and sys.version_info < (
+        3,
+        8,
+        0,
+    ):
```
Comment (contramundum53) [inline tests/integration_tests/test_botorch.py:109]: Note: `botorch >=0.8.1` is not available in python 3.7, so I had to skip some tests.

File: optuna/integration/botorch.py [hunk 2]
```diff
@@ -63,6 +63,100 @@ def _get_sobol_qmc_normal_sampler(num_samples: int) -> SobolQMCNormalSampler:
...
+    Returns:
+        Next set of candidates. Usually the return value of BoTorch's ``optimize_acqf``.
+
+    """
+
+    # We need botorch >=0.8.1 for LogExpectedImprovement.
+    if not _imports_logei.is_successful():
+        raise ImportError(
+            "logei_candidates_func requires botorch >=0.8.1. "
+            "Please upgrade botorch or use qei_candidates_func instead."
```
Comment (not522) [inline optuna/integration/botorch.py:117]: How about clarifying that they are used as `candidates_func`?

```suggestion
            "Please upgrade botorch or use qei_candidates_func as candidates_func instead."
```

**Review by not522 — COMMENTED** [review]
Thank you for your PR. Could you add `logei_candidates_func` in this list? https://github.com/optuna/optuna/blob/master/docs/source/reference/integration.rst#botorch

**Review by not522 — COMMENTED** [review]
LGTM except for the version.

**Review by not522 — APPROVED** [review]

File: optuna/integration/botorch.py [hunk 3]
```diff
@@ -63,6 +63,100 @@ def _get_sobol_qmc_normal_sampler(num_samples: int) -> SobolQMCNormalSampler:
...
+    train_x = normalize(train_x, bounds=bounds)
+
+    model = SingleTaskGP(train_x, train_y, outcome_transform=Standardize(m=train_y.size(-1)))
+    mll = ExactMarginalLogLikelihood(model.likelihood, model)
+    fit_gpytorch_mll(mll)
+
+    acqf = LogExpectedImprovement(
+        model=model,
+        best_f=best_f,
+        sampler=_get_sobol_qmc_normal_sampler(256),
```
Comment (toshihikoyanase) [inline optuna/integration/botorch.py:140]: I cannot find the `sampler` argument in the `LogExpectedImprovement` reference. Since it employs an analytical approach according to the module name, it might not require it.

```suggestion
```

File: tests/integration_tests/test_botorch.py [hunk 4]
```diff
@@ -32,6 +33,9 @@
 
 @pytest.mark.parametrize("n_objectives", [1, 2, 4])
 def test_botorch_candidates_func_none(n_objectives: int) -> None:
+    if n_objectives == 1 and sys.version_info < (3, 8):
```
Comment (toshihikoyanase) [inline tests/integration_tests/test_botorch.py:36]: The logic used to bypass the tests seems somewhat indirect. It could be clearer if it directly verifies the version of botorch, in a similar manner to what's demonstrated in `optuna/integration/botorch.py`. Would that be a feasible approach for you to consider?

https://github.com/optuna/optuna/blob/2b6addc365003eaab9e20bacb964ff15ed057da7/optuna/integration/botorch.py#L42

**Review by toshihikoyanase — COMMENTED** [review]
Sorry for the delayed response. The change overall looks great to me, but I have some small comments.

File: optuna/integration/botorch.py [hunk 5]
```diff
@@ -63,6 +63,100 @@ def _get_sobol_qmc_normal_sampler(num_samples: int) -> SobolQMCNormalSampler:
...
+            and ``n_params`` is the number of parameters. ``n_params`` may be larger than the
+            actual number of parameters if categorical parameters are included in the search
+            space, since these parameters are one-hot encoded.
+            Values are not normalized.
+        train_obj:
+            Previously observed objectives. A ``torch.Tensor`` of shape
+            ``(n_trials, n_objectives)``. ``n_trials`` is identical to that of ``train_x``.
+            ``n_objectives`` is the number of objectives. Observations are not normalized.
+        train_con:
+            Objective constraints. This option is not supported in logei_candidates_func and
```
Comment (toshihikoyanase) [inline optuna/integration/botorch.py:96]: A nit.
```suggestion
            Objective constraints. This option is not supported in ``logei_candidates_func`` and
```

File: optuna/integration/botorch.py [hunk 6]
```diff
@@ -63,6 +63,100 @@ def _get_sobol_qmc_normal_sampler(num_samples: int) -> SobolQMCNormalSampler:
...
+    train_x: "torch.Tensor",
+    train_obj: "torch.Tensor",
+    train_con: Optional["torch.Tensor"],
+    bounds: "torch.Tensor",
+    pending_x: Optional["torch.Tensor"],
+) -> "torch.Tensor":
+    """Log Expected Improvement (LogEI).
+
+    The default value of ``candidates_func`` in :class:`~optuna.integration.BoTorchSampler`
+    with single-objective optimization for non-constrained problems.
```
Comment (toshihikoyanase) [inline optuna/integration/botorch.py:81]: We can update the description of the default `candidates_func` in the `BoTorchSampler` reference as well.

https://github.com/optuna/optuna/blob/2b6addc365003eaab9e20bacb964ff15ed057da7/optuna/integration/botorch.py#L462-L468

File: optuna/integration/botorch.py (hunk 3 above)
Comment (contramundum53) [reply optuna/integration/botorch.py:140]: Nice catch! Thaks for the comment.

**Comment by contramundum53** [comment]
@toshihikoyanase  I applied all your comments. PTAL.

**Comment by toshihikoyanase** [comment]
@contramundum53 Thank you for your update. The `checks` CI did not pass due to some extra spaces at the end of the lines. You can find the details at https://github.com/optuna/optuna/actions/runs/5318669226/jobs/9630367164.

File: optuna/integration/botorch.py [hunk 7]
```diff
@@ -467,8 +564,11 @@ class BoTorchSampler(BaseSampler):
             :obj:`None`. For any constraints that failed to compute, the tensor will contain
             NaN.
 
-            If omitted, it is determined automatically based on the number of objectives. If the
-            number of objectives is one, Quasi MC-based batch Expected Improvement (qEI) is used.
+            If omitted, it is determined automatically based on the number of objectives and 
+            whether a constraint is specified. If the
+            number of objectives is one and no constraint is specified, log-Expected Improvement
+            is used. If constraints are specified, quasi MC-based batch Expected Improvement
```
Comment (contramundum53) [inline optuna/integration/botorch.py:570]:
```suggestion
            is used. If constraints are specified, quasi MC-based batch Expected Improvement
```

**Comment by contramundum53** [comment]
@toshihikoyanase PTAL.

File: optuna/integration/botorch.py (hunk 6 above)
Comment (toshihikoyanase) [reply optuna/integration/botorch.py:81]: Thanks for your update.

File: tests/integration_tests/test_botorch.py (hunk 1 above)
Comment (toshihikoyanase) [reply tests/integration_tests/test_botorch.py:109]: Skipping the test based on the Python version seems to be a bit indirect. How about checking the `botorch` version instead?

**Review by toshihikoyanase — COMMENTED** [review]
Thanks for your update. The change in the sampler looks solid, but I have a comment on the tests. https://github.com/optuna/optuna/pull/4667#discussion_r1246029780

File: tests/integration_tests/test_botorch.py (hunk 1 above)
Comment (contramundum53) [reply tests/integration_tests/test_botorch.py:109]:
```suggestion
    if candidates_func == integration.botorch.logei_candidates_func and version.parse(botorch.version.version) < version.parse("0.8.1"):
```

File: tests/integration_tests/test_botorch.py [hunk 8]
```diff
@@ -97,6 +103,11 @@ def candidates_func(
     ],
 )
 def test_botorch_specify_candidates_func(candidates_func: Any, n_objectives: int) -> None:
+    if candidates_func == integration.botorch.logei_candidates_func and version.parse(
+        botorch.version.version
+    ) < version.parse("0.8.1"):
+        pytest.skip("LogExpectedImprovement is not available in Python <3.8.")
```
Comment (toshihikoyanase) [inline tests/integration_tests/test_botorch.py:109]:
```suggestion
        pytest.skip("LogExpectedImprovement is not available in botorch <0.8.1.")
```

File: tests/integration_tests/test_botorch.py [hunk 9]
```diff
@@ -32,6 +34,9 @@
 
 @pytest.mark.parametrize("n_objectives", [1, 2, 4])
 def test_botorch_candidates_func_none(n_objectives: int) -> None:
+    if n_objectives == 1 and version.parse(botorch.version.version) < version.parse("0.8.1"):
+        pytest.skip("Python >=3.8 is required for logei_candidates_func.")
```
Comment (toshihikoyanase) [inline tests/integration_tests/test_botorch.py:38]:
```suggestion
        pytest.skip("botorch >=0.8.1 is required for logei_candidates_func.")
```

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Could you update the messages as well?

**Review by toshihikoyanase — APPROVED** [review]
Thanks for your update. LGTM!
